File: example3/main.cpp
#include <iostream>
#include <opencv2/opencv.hpp>

class Context;

class State
{
protected:
    Context *context_;

public:
    virtual ~State() {}

    void set_context(Context *context)
    {
        this->context_ = context;
    }

    virtual void handleRequest1() = 0;
    virtual void handleRequest2() = 0;
    virtual void update() = 0;
};

class Context
{
private:
    State *state_;

public:
    cv::Mat image;

    Context(State *state, cv::Mat image) : state_(nullptr)
    {
        this->TransitionTo(state);
        this->image = image;
    }

    ~Context()
    {
        delete state_;
    }

    void TransitionTo(State *state)
    {
        if (this->state_ != nullptr)
            delete this->state_;
        this->state_ = state;
        this->state_->set_context(this);
    }

    void handleRequest1()
    {
        this->state_->handleRequest1();
    }

    void handleRequest2()
    {
        this->state_->handleRequest2();
    }

    void update()
    {
        this->state_->update();
    }
};

class HandleState1 : public State
{
public:
    void handleRequest1() override
    {
        std::cout << "Handle1.\n";
        
    }

    void handleRequest2() override;

    void update() override
    {
        cv::putText(this->context_->image, "Handle1", cv::Point(10, 50), cv::FONT_HERSHEY_SIMPLEX, 1, cv::Scalar(0, 0, 255), 2);
        cv::imshow("Image", this->context_->image);
    }
};

class HandleState2 : public State
{
public:
    void handleRequest1() override
    {
        std::cout << "Transition to Handle1.\n";
        this->context_->TransitionTo(new HandleState1);
    }

    void handleRequest2() override
    {
        std::cout << "Handle2.\n";
        
    }

    void update() override
    {
        cv::putText(this->context_->image, "Handle2", cv::Point(10, 50), cv::FONT_HERSHEY_SIMPLEX, 1, cv::Scalar(255, 0, 0), 2);
        cv::imshow("Image", this->context_->image);
    }
};

void HandleState1::handleRequest2()
{
    std::cout << "Transition to Handle2.\n";
    this->context_->TransitionTo(new HandleState2);
}

int main()
{
    cv::Mat image = cv::imread("../../example3/img/image.png");
    Context *context = new Context(new HandleState1, image);
    context->handleRequest1();
    while (true)
    {
        context->image = cv::imread("../../example3/img/image.png");
        context->update();
        char c = (char)cv::waitKey(1);
        if (c == 27)
            break;
        else if (c == '1')
            context->handleRequest1();
        else if (c == '2')
            context->handleRequest2();
    }
    return 0;
}

File: example3/main.py
from __future__ import annotations
import cv2
from abc import ABC, abstractmethod

class Context:
    _state = None
    handleNumber = None
    img = None
    color = None

    def __init__(self, state: State) -> None:
        self.transition_to(state)

    def transition_to(self, state: State):
        self._state = state
        self._state.context = self

    def handleRequest1(self):
        self._state.handle1()
    
    def handleRequest2(self):
        self._state.handle2()
    
    def update(self):
        self._state.update()

class State(ABC):
    @property
    def context(self) -> Context:
        return self._context
    
    @context.setter
    def context(self, context: Context) -> None:
        self._context = context
    
    @abstractmethod
    def handle1(self) -> None:
        pass
    
    @abstractmethod
    def handle2(self) -> None:
        pass

    @abstractmethod
    def update(self) -> None:
        cv2.putText(self._context.img, f"{self._context.handleNumber}", (10, 30), cv2.FONT_HERSHEY_SIMPLEX, 1, self._context.color, 2)
        cv2.imshow("Img", self._context.img)

class HandleState1(State):
    def handle1(self) -> None:
        print("Handle State 1")

    def handle2(self) -> None:
        print("Transition to Handle State 2")
        self.context.transition_to(HandleState2())

    def update(self) -> None:
        self.context.color = (0, 0, 255)
        self.context.handleNumber = "Handle1"
        return super().update()
    
class HandleState2(State):
    def handle1(self) -> None:
        print("Transition to Handle State 1")
        self.context.transition_to(HandleState1())
    
    def handle2(self) -> None:
        print("Handle State 2")

    def update(self) -> None:
        self.context.color = (255, 0, 0)
        self.context.handleNumber = "Handle2"
        return super().update()

if __name__ == "__main__":
    context = Context(HandleState1())
    context.transition_to(HandleState1())
    while True:
        context.img = cv2.imread("./example3/img/image.png")
        context.update()
        key = cv2.waitKey(1) & 0xFF
        if key == ord("q"):
            break
        elif key == ord("1"):
            context.handleRequest1()
        elif key == ord("2"):
            context.handleRequest2()
    cv2.destroyAllWindows()

    